fendermint/vm/interpreter/src/fvm/state/snapshot.rs

```diff
@@ -0,0 +1,312 @@
+use std::path::Path;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use tokio_util::compat::TokioAsyncWriteCompatExt;
+
+pub type BlockHeight = u64;
+
+/// Taking snapshot of the current blockchain state
+pub enum Snapshot<DB> {
+    V1(V1Snapshot<DB>),
+}
+
+impl<DB> Snapshot<DB>
+where
+    DB: Blockstore + Clone + 'static + Send,
+{
+    pub fn new(
+        store: DB,
+        state_params: FvmStateParams,
+        block_height: BlockHeight,
+    ) -> anyhow::Result<Self> {
+        Ok(Self::V1(V1Snapshot::new(
+            store,
+            state_params,
+            block_height,
+        )?))
+    }
+
+    pub fn version(&self) -> u64 {
+        match self {
+            Snapshot::V1(_) => 1,
+        }
+    }
+
+    /// Init the block chain from the snapshot
+    pub fn init_chain(&self) -> anyhow::Result<()> {
+        todo!()
+    }
+
+    /// Read the snapshot from file
+    pub async fn read_car(_path: impl AsRef<Path>) -> anyhow::Result<Self> {
+        todo!()
+    }
+
+    /// Write the snapshot to car file
+    pub async fn write_car(self, path: impl AsRef<Path>) -> anyhow::Result<()> {
+        let file = tokio::fs::File::create(path).await?;
+
+        // derive the car header roots
+        let car = CarHeader::new(self.car_header_roots()?, self.version());
+
+        match self {
+            Snapshot::V1(s) => s.write_car(car, file).await?,
+        }
+
+        Ok(())
+    }
+
+    fn car_header_roots(&self) -> anyhow::Result<Vec<Cid>> {
+        match self {
+            Snapshot::V1(s) => s.car_header_roots(),
+        }
+    }
+}
+
+pub struct V1Snapshot<DB> {
+    /// The root cid of the state tree
+    state_tree_root: Cid,
+    /// The state tree of the current blockchain
+    state_tree: StateTree<ReadOnlyBlockstore<DB>>,
+    /// The latest block height with state params serialized to bytes
+    block_state_params_bytes: Vec<u8>,
+    /// Block state params cid derived from block state params bytes
+    block_state_params_cid: Cid,
+}
+
+impl<DB> V1Snapshot<DB>
+where
+    DB: Blockstore + Clone + 'static + Send,
+{
+    /// Creates a new V2Snapshot struct. Caller ensure store
+    pub fn new(
+        store: DB,
+        state_params: FvmStateParams,
+        block_height: BlockHeight,
+    ) -> anyhow::Result<Self> {
+        let state_tree =
+            StateTree::new_from_root(ReadOnlyBlockstore::new(store), &state_params.state_root)?;
+        let state_tree_root = state_params.state_root;
+
+        let block_state_params_bytes = fvm_ipld_encoding::to_vec(&(state_params, block_height))?;
+        let block_state_params_cid =
+            Cid::new_v1(DAG_CBOR, Code::Blake2b256.digest(&block_state_params_bytes));
+
+        Ok(Self {
+            state_tree_root,
+            state_tree,
+            block_state_params_bytes,
+            block_state_params_cid,
+        })
+    }
+
+    /// For V1 snapshot, we are putting two components into the CAR file: the state tree and latest
```

Comment:
I don't know if it makes sense in Fendermint (or if we even need them) because maybe Tendermint already includes a way to recover from a snapshot or genesis, but in Lotus snapshots include also all transactions from genesis (or a previous snapshot) so one could verify if the state tree is correct by re-executing all messages.

We should maybe consider adding the option to include messages in a snapshot?

Reply:
We wouldn't have access to all those messages unless we store them or query them from Tendermint, but I don't think this is the intended use of snapshots in Tendermint, since it can garbage collect old blocks.

Reply:
Amazing, then. If Tendermint handles this for us then no need to include messages in snapshots. So here what we are building is actually a "BFT-snapshot", i.e. the baseline state required for a node to sync from a height.